0.3.31: (shared) build failures with USE_THREAD=0 USE_OPENMP=0

First is `linktest.c:(.text.startup+0x96f): undefined reference to `openblas_set_threads_callback_function'`:

```
./gensymbol linktest  x86_64 "_" 1 0 1 1 0 0 "" "" 0 0 0 1 1 1 1 > linktest.c
x86_64-pc-linux-gnu-cc -march=native -O2 -pipe -O2 -DMAX_STACK_ALLOC=2048 -DEXPRECISION -m128bit-long-double -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.31\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -I. -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DEXPRECISION -m128bit-long-double -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.31\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -I.. -Wl,-O1 -Wl,--as-needed -shared -o ../libopenblas_haswell-r0.3.31.so \
-Wl,--whole-archive ../libopenblas_haswell-r0.3.31.a -Wl,--no-whole-archive \
-Wl,-soname,libopenblas.so.0 -lm -lm
x86_64-pc-linux-gnu-cc -march=native -O2 -pipe -O2 -DMAX_STACK_ALLOC=2048 -DEXPRECISION -m128bit-long-double -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.31\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -I. -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -DEXPRECISION -m128bit-long-double -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.31\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -I.. -Wl,-O1 -Wl,--as-needed -w -o linktest linktest.c ../libopenblas_haswell-r0.3.31.so -L/usr/x86_64-pc-linux-gnu/lib/gcc/x86_64-pc-linux-gnu/15 -L/usr/x86_64-pc-linux-gnu/lib/gcc/x86_64-pc-linux-gnu/15/../../../../x86_64-pc-linux-gnu/lib -L/usr/x86_64-pc-linux-gnu/lib/gcc/x86_64-pc-linux-gnu/15/../../.. -L/usr/x86_64-pc-linux-gnu/lib  -lgfortran -lm -lquadmath -lm -lc  && echo OK.
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: /var/tmp/paludis/build/sci-libs-OpenBLAS-0.3.31/temp//ccQbPhYL.o: in function `main':
linktest.c:(.text.startup+0x96f): undefined reference to `openblas_set_threads_callback_function'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:211: ../libopenblas_haswell-r0.3.31.so] Error 1
make[1]: Leaving directory '/var/tmp/paludis/build/sci-libs-OpenBLAS-0.3.31/work/OpenBLAS-0.3.31/exports'
make: *** [Makefile:149: shared] Error 2
```

complete build log: [OpenBLAS-0.3.31_01_build.log](https://github.com/user-attachments/files/24670903/OpenBLAS-0.3.31_01_build.log)

Which can be worked around by reverting 879497990f42e6dcfa370bcbf478cc5674158336

And then hits `test_post_fork_async.c:(.text+0xf8): undefined reference to `dgetrf_'`:

```
x86_64-pc-linux-gnu-cc -march=native -O2 -pipe -O2 -DMAX_STACK_ALLOC=2048 -DEXPRECISION -m128bit-long-double -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.31\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -I. -O2 -DSMALL_MATRIX_OPT -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=f_ -DMAX_STACK_ALLOC=2048 -DEXPRECISION -m128bit-long-double -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.31\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -I.. -Wl,-O1 -Wl,--as-needed -o openblas_utest utest_main.o test_min.o test_amax.o test_ismin.o test_rotmg.o test_axpy.o test_dotu.o test_dsdot.o test_swap.o test_rot.o test_dnrm2.o test_zscal.o test_amin.o test_axpby.o test_gemv.o test_fork.o test_post_fork_async.o test_post_fork.o ../libopenblas_haswell-r0.3.31.a -lm -lm -L/usr/x86_64-pc-linux-gnu/lib/gcc/x86_64-pc-linux-gnu/15 -L/usr/x86_64-pc-linux-gnu/lib/gcc/x86_64-pc-linux-gnu/15/../../../../x86_64-pc-linux-gnu/lib -L/usr/x86_64-pc-linux-gnu/lib/gcc/x86_64-pc-linux-gnu/15/../../.. -L/usr/x86_64-pc-linux-gnu/lib  -lgfortran -lm -lquadmath -lm -lc 
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: test_post_fork_async.o: in function `__ctest_fork_safety_after_fork_async_run':
test_post_fork_async.c:(.text+0xf8): undefined reference to `dgetrf_'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:78: openblas_utest] Error 1
make[1]: Leaving directory '/var/tmp/paludis/build/sci-libs-OpenBLAS-0.3.31/work/OpenBLAS-0.3.31/utest'
make: *** [Makefile:177: tests] Error 2
```

complete build log: [OpenBLAS-0.3.31_02_build.log](https://github.com/user-attachments/files/24670910/OpenBLAS-0.3.31_02_build.log)

Which can be worked around by reverting  7750d507ff7e34ce354ca66954be447b50d313b2

--- a/driver/others/CMakeLists.txt
+++ b/driver/others/CMakeLists.txt
@@ -27,7 +27,6 @@ if (USE_THREAD)
     ${BLAS_SERVER}
     divtable.c # TODO: Makefile has -UDOUBLE
     blas_l1_thread.c
-    blas_server_callback.c
   )
 
   if (NOT NO_AFFINITY)
@@ -42,6 +41,7 @@ set(COMMON_SOURCES
   openblas_env.c
   openblas_get_num_procs.c
   openblas_get_num_threads.c
+  blas_server_callback.c
 )
 
 # these need to have NAME/CNAME set, so use GenerateNamedObjects, but don't use standard name mangling

--- a/driver/others/Makefile
+++ b/driver/others/Makefile
@@ -1,12 +1,12 @@
 TOPDIR	= ../..
 include ../../Makefile.system
 
-COMMONOBJS	 = memory.$(SUFFIX) xerbla.$(SUFFIX) c_abs.$(SUFFIX) z_abs.$(SUFFIX) openblas_set_num_threads.$(SUFFIX) openblas_get_num_threads.$(SUFFIX) openblas_get_num_procs.$(SUFFIX) openblas_get_config.$(SUFFIX) openblas_get_parallel.$(SUFFIX) openblas_error_handle.$(SUFFIX) openblas_env.$(SUFFIX)
+COMMONOBJS	 = memory.$(SUFFIX) xerbla.$(SUFFIX) c_abs.$(SUFFIX) z_abs.$(SUFFIX) openblas_set_num_threads.$(SUFFIX) openblas_get_num_threads.$(SUFFIX) openblas_get_num_procs.$(SUFFIX) openblas_get_config.$(SUFFIX) openblas_get_parallel.$(SUFFIX) openblas_error_handle.$(SUFFIX) openblas_env.$(SUFFIX) blas_server_callback.$(SUFFIX)
 
 #COMMONOBJS	+= slamch.$(SUFFIX) slamc3.$(SUFFIX) dlamch.$(SUFFIX)  dlamc3.$(SUFFIX)
 
 ifdef SMP
-COMMONOBJS	+= blas_server.$(SUFFIX) divtable.$(SUFFIX) blasL1thread.$(SUFFIX) blas_server_callback.$(SUFFIX)
+COMMONOBJS	+= blas_server.$(SUFFIX) divtable.$(SUFFIX) blasL1thread.$(SUFFIX)
 ifneq ($(NO_AFFINITY), 1)
 COMMONOBJS	+= init.$(SUFFIX)
 endif

--- a/utest/CMakeLists.txt
+++ b/utest/CMakeLists.txt
@@ -101,8 +101,13 @@ if (NOT USE_OPENMP)
 set(OpenBLAS_utest_src
   ${OpenBLAS_utest_src}
   test_fork.c
+  )
+if (NOT NO_LAPACK)
+set(OpenBLAS_utest_src
+  ${OpenBLAS_utest_src}
   test_post_fork_async.c
   )
+endif()
 endif()
 set(OpenBLAS_utest_src
   ${OpenBLAS_utest_src}

--- a/utest/Makefile
+++ b/utest/Makefile
@@ -45,7 +45,10 @@ endif
 # FIXME TBD if this works on OSX, SunOS, POWER and zarch
 ifeq ($(OSNAME), $(filter $(OSNAME),Linux CYGWIN_NT))
 ifneq ($(USE_OPENMP), 1)
-OBJS += test_fork.o test_post_fork_async.o
+OBJS += test_fork.o
+ifneq ($(NO_LAPACK), 1)
+OBJS += test_post_fork_async.o
+endif
 endif
 OBJS += test_post_fork.o
 endif

